src/nvim/eval.c

```diff
@@ -15577,6 +15577,53 @@ static void f_split(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   p_cpo = save_cpo;
 }
 
+/// "stdpath()" helper for list results
+void stdpath_list_helper(const XDGVarType xdg, typval_T *rettv)
+{
+  const void *iter = NULL;
+  list_T *const list = tv_list_alloc();
+  rettv->v_type = VAR_LIST;
+  rettv->vval.v_list = list;
+  list->lv_refcount++;
+  char *const dirs = stdpaths_get_xdg_var(xdg);
+  do {
+    size_t dir_len;
+    const char *dir;
+    iter = vim_colon_env_iter(dirs, iter, &dir, &dir_len);
+    if (dir != NULL && dir_len > 0) {
+      char *dir_with_nvim = xmemdupz(dir, dir_len);
+      dir_with_nvim = concat_fnames_realloc(dir_with_nvim, "nvim", true);
+      tv_list_append_string(list, dir_with_nvim, strlen(dir_with_nvim));
+      xfree(dir_with_nvim);
+    }
+  } while (iter != NULL);
+  xfree(dirs);
+}
+
+/// "stdpath(type)" function
+static void f_stdpath(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+{
+  const char *const p = tv_get_string_chk(&argvars[0]);
+  if (p == NULL) {
+    return;  // Type error; errmsg already given.
+  }
+  rettv->v_type = VAR_STRING;
+
+  if (strcmp(p, "config") == 0) {
+    rettv->vval.v_string = (char_u *)get_xdg_home(kXDGConfigHome);
+  } else if (strcmp(p, "data") == 0) {
+    rettv->vval.v_string = (char_u *)get_xdg_home(kXDGDataHome);
+  } else if (strcmp(p, "cache") == 0) {
+    rettv->vval.v_string = (char_u *)get_xdg_home(kXDGCacheHome);
+  } else if (strcmp(p, "config_dirs") == 0) {
+    stdpath_list_helper(kXDGConfigDirs, rettv);
+  } else if (strcmp(p, "data_dirs") == 0) {
+    stdpath_list_helper(kXDGDataDirs, rettv);
+  } else {
+    EMSG(_(e_invarg));
```

Comment:
Never use `e_invarg` in new code: it does not provide any details, even which function received invalid argument, and cannot be tagged in documentation. Need new (four-digit, far from Vim ones) error code and a relevant tag in `stdpath()` documentation.

Also that does not free you from assigning something to `vval.v_string`, given that pointers are 64-bit and varnumber_T is 32-bit (by default VAR_NUMBER zero is assigned) I do not know what string may point to.